samples/js/flickr/slurp/src/main.js

```diff
@@ -58,44 +62,72 @@ async function main(): Promise<void> {
+  const photosets = getPhotosets();
 
-  const photosets = await getPhotosets(flickr);
-  let seen = 0;
-  const photosetsPromise = photosets.map(p => getPhotoset(flickr, p.id).then(p => {
-    process.stdout.write(`${clearLine}${++seen} of ${photosets.length} photosets imported...`);
-    return jsonToNoms(p);
-  }));
-  const setOfPhotosets = new Set(await Promise.all(photosetsPromise));
-
-  process.stdout.write(clearLine);
-  return db.commit(out, newStruct('', {
-    photosetsMeta: jsonToNoms(photosets),
-    photosets: await setOfPhotosets,
+  await db.commit(out, jsonToNoms({
+    photos: await photos,
+    photosets: await photosets,
   }), {
     meta: newStruct('', {date: new Date().toISOString()}),
-  })
-  .then(() => db.close());
+  });
+  await db.close();
+  process.stdout.write(clearLine);
+}
+
+function getUserId(): Promise<string> {
+  return flickr.callApi('flickr.urls.getUserProfile').then(d => d.user.nsid);
 }
 
-async function getPhotoset(flickr: Flickr, id: string): Promise<*> {
-  const json = await flickr.callApi('flickr.photosets.getPhotos', {
-    'photoset_id': id,
-    extras: 'license, date_upload, date_taken, owner_name, icon_server, original_format, ' +
-      'last_update, geo, tags, machine_tags, o_dims, views, media, path_alias, url_sq, url_t, ' +
-      'url_s, url_m, url_o',
+function getAllPhotos(userId: string): Promise<Array<any>> {
+  const minPrivacy = 1;
+  const maxPrivacy = 5;
+  const results = [];
+  for (let i = minPrivacy; i <= maxPrivacy; i++) {
+    results.push(getPhotos(userId, i));
+  }
+  return Promise.all(results);
+}
+
+async function getPhotos(userId: string, privacyFilter: number): Promise<Array<any>> {
+  const perPage = 500;
+  const p1 = await getPhotoPage(userId, privacyFilter, perPage, 1);
+  totalPhotos += Number(p1.photos.total);
+  updateStatus();
+  const results = [];
+  for (let i = 2; i <= p1.photos.pages; i++) {
+    results.push(getPhotoPage(userId, privacyFilter, perPage, i));
+  }
+  return [p1].concat(await Promise.all(results));
+}
+
+async function getPhotoPage(userId: string, privacyFilter: number, perPage: number, pageNumber: number): Promise<any> {
+  const result = await flickr.callApi('flickr.photos.search', {
+    user_id: userId,
+    privacy_filter: String(privacyFilter),
+    per_page: String(perPage),
+    page: String(pageNumber),
+    extras: 'description,license,date_upload,date_taken,owner_name,icon_server,original_format,last_update,geo,tags,machine_tags,o_dims,views,media,path_alias,url_sq,url_t,url_s,url_q,url_m,url_n,url_z,url_c,url_l,url_o',
   });
-  return json.photoset;
+  gottenPhotos += Number(result.photos.photo.length);
+  updateStatus();
+  return result;
+}
+
+async function getPhotosets(): Promise<any> {
+  return await flickr.callApi('flickr.photosets.getList');
 }
 
-function getPhotosets(flickr: Flickr): Promise<*> {
-  return flickr.callApi('flickr.photosets.getList').then(v => v.photosets.photoset);
+function updateStatus() {
```

Comment:
Why don't you pass in `gottenPhotos` and `totalPhotos`?

Another option is to wrap these in a class with local state.